mountpoint-s3-crt/src/s3/pool.rs

```diff
@@ -0,0 +1,429 @@
+            );
+        }
+
+        ticket
+    }
+}
+
+unsafe extern "C" fn pool_reserve<Pool: MemoryPool>(
+    pool: *mut aws_s3_buffer_pool,
+    meta: aws_s3_buffer_pool_reserve_meta,
+) -> *mut aws_future_s3_buffer_ticket {
+    // SAFETY: `pool` was obtained through `CrtMemoryPool::leak`.
+    let crt_pool = unsafe { CrtBufferPool::<Pool>::ref_from_raw(&pool) };
+    let future = crt_pool.reserve(meta.size);
+
+    // SAFETY: the CRT will take ownership of the future.
+    unsafe { future.into_inner_ptr() }
+}
+
+unsafe extern "C" fn pool_trim<Pool: MemoryPool>(pool: *mut aws_s3_buffer_pool) {
+    // SAFETY: `pool` was obtained through `CrtMemoryPool::leak`.
+    let crt_pool = unsafe { CrtBufferPool::<Pool>::ref_from_raw(&pool) };
+    crt_pool.trim();
+}
+
+unsafe extern "C" fn pool_destroy<Pool: MemoryPool>(data: *mut libc::c_void) {
+    let pool = data as *mut aws_s3_buffer_pool;
+
+    // SAFETY: `pool` was obtained through `CrtMemoryPool::leak`.
+    _ = unsafe { CrtBufferPool::<Pool>::from_raw(pool) };
+}
+
+/// Wrapper for [aws_s3_buffer_ticket].
+struct CrtTicket<Buffer: AsMut<[u8]>> {
+    /// Inner struct to pass to CRT functions.
+    inner: aws_s3_buffer_ticket,
+    /// Buffer implementing [AsMut<\[u8\]>].
+    buffer: Buffer,
+    /// Pin this struct because inner.impl_ will be a pointer to this object.
+    _pinned: PhantomPinned,
+}
+
+impl<Buffer: AsMut<[u8]>> CrtTicket<Buffer> {
+    /// Leak a pinned instance and returns a raw pointer.
+    ///
+    /// # Safety
+    /// The returned pointer must eventually be passed to [from_raw] and can
+    /// additionally only used in [ref_mut_from_raw].
+    unsafe fn leak(self: Pin<Box<Self>>) -> *mut aws_s3_buffer_ticket {
+        // SAFETY: the resulting pointer will be only used in `ticket_claim` and `ticket_destroy`.
+        let boxed = unsafe { Pin::into_inner_unchecked(self) };
+        let pool = Box::leak(boxed);
+        &raw mut pool.inner
+    }
+
+    /// Returns a reference to original instance from a raw pointer.
+    ///
+    /// # Safety
+    /// The raw pointer must have been obtained through [leak()].
+    unsafe fn ref_mut_from_raw(ticket: &mut *mut aws_s3_buffer_ticket) -> &mut Self {
+        // SAFETY: `ticket` points to the `inner` field of a pinned instance.
+        unsafe {
+            let impl_ptr = (**ticket).impl_;
+            &mut *(impl_ptr as *mut Self)
+        }
+    }
+
+    /// Re-constructs the original pinned instance from a raw pointer.
+    ///
+    /// # Safety
+    /// The raw pointer must have been obtained through [leak()].
+    unsafe fn from_raw(ticket: *mut aws_s3_buffer_ticket) -> Pin<Box<Self>> {
+        // SAFETY: `ticket` points to the `inner` field of a pinned instance.
+        unsafe { Pin::new_unchecked(Box::from_raw((*ticket).impl_ as *mut Self)) }
+    }
+}
+
+unsafe extern "C" fn ticket_claim<Buffer: AsMut<[u8]>>(mut ticket: *mut aws_s3_buffer_ticket) -> aws_byte_buf {
```

Comment:
Could we have some docs on what this function is meant to do? Does it convert the buffer inside `aws_s3_buffer_ticket` into an `aws_byte_buf`?